### ISSUE TITLE
🐛 BUG: hoisted scripts in Astro's components make the compiler goes crazy in `astro@next`

### ISSUE THREAD

**Issue by Chrissdroid** [snowpackjs/astro#1741]
### What package manager are you using?

npm

### What operating system are you using?

Linux

### Describe the Bug

The compiler throw an "Unterminated string literal" error in the console at a mandatory line which is not at the script hoist position and not affected by the file content, which is very confusing to know what's going on for debugging

also generating hundreds of NL characters in the output log

### Temporal solution for now

just remove the hoist attribute on the script

### Steps to Reproduce

1. `npm init astro` using template [Astro Minimal exemple](https://astro.new/minimal@next?on=stackblitz)
2. add script tag with hoist attribute
3. See the compiller throws "Unterminated string literal" at you at a wrong error line position

### Link to Minimal Reproducible Example (Optional)

https://stackblitz.com/edit/github-zlglc8?devtoolsheight=33&file=src/pages/index.astro

**Comment by aFuzzyBear** [comment]
Please note that this was produced on the `astro@next` beta `v0.21`

**Comment by matthewp** [comment]
Thanks!

**Comment by matthewp** [comment]
Ok, I can confirm something weird is going on with hoisted scripts. This is surprising as we have tests here: https://github.com/snowpackjs/astro-compiler-next/blob/main/internal/printer/printer_test.go

However the tests in Astro are still commented out: https://github.com/snowpackjs/astro/blob/d84bfe719a546ad855640338d5ed49ad3aa4ccb4/packages/astro/test/astro-scripts.test.js

And I can confirm that they are now failing. So I created a PR to add the tests back: https://github.com/snowpackjs/astro/pull/1743

**Comment by borisv** [comment]
Could reproduce this across the board of templates. It throws
```
Error: Transform failed with 1 error:
/root/_DEV/astro.next/minimal/src/pages/index.astro:16:135: error: Unterminated string literal
    at failureErrorWithLog (/root/_DEV/astro.next/minimal/node_modules/esbuild/lib/main.js:1493:15)
    at /root/_DEV/astro.next/minimal/node_modules/esbuild/lib/main.js:1282:29
    at /root/_DEV/astro.next/minimal/node_modules/esbuild/lib/main.js:629:9
    at handleIncomingPacket (/root/_DEV/astro.next/minimal/node_modules/esbuild/lib/main.js:726:9)
    at Socket.readFromStdout (/root/_DEV/astro.next/minimal/node_modules/esbuild/lib/main.js:596:7)
    at Socket.emit (events.js:315:20)
    at addChunk (internal/streams/readable.js:309:12)
    at readableAddChunk (internal/streams/readable.js:284:9)
    at Socket.Readable.push (internal/streams/readable.js:223:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:188:23)
09:00 PM [astro] 500 /                                        5ms
```
Where the only thing changing between different templates is the erratic line and column numbers that are thrown, while the numbers stay the same within any template no matter where the script tag is positioned.

I remember having seen such unrelated line numbers thrown quite often in an early stage of the new go compiler development. 

Tested with minimal, blog, starter, docs using astro@0.21.0-next.1 beta v0.21

**Comment by borisv** [comment]
There is a difference between when the script tag contains code and when not. There is no detailed error message when the tag contains code only throwing 500 in the terminal and

```
500 Internal Error
/root/_DEV/astro.next/docs/src/pages/index.astro: Unterminated string literal
```
in the browser.

Also in the  latter case the 500 comes only after a browser refresh.

**Comment by natemoo-re** [comment]
Tracked this down! See https://github.com/snowpackjs/astro-compiler-next/pull/138.

TLDR: if a component only contained `<script hoist>`, when we stringified our AST we weren't being smart enough to handle an empty document (since all the children have been hoisted out).

**Comment by natemoo-re** [comment]
Following up on my previous comment, the team has decided to close out issues that have been confirmed as fixed by [astro@0.21.0-next.0](https://github.com/snowpackjs/astro/releases/tag/astro%400.21.0-next.0), [astro@0.21.0-next.1](https://github.com/snowpackjs/astro/releases/tag/astro%400.21.0-next.1), or [astro@0.21.0-next.2](https://github.com/snowpackjs/astro/releases/tag/astro%400.21.0-next.2). Our hope is that this will help the [v0.21 milestone](https://github.com/snowpackjs/astro/milestone/1) remain as actionable as possible.

To verify that this issue has been fixed, you may
- Open a new [`astro@next` project on StackBlitz](https://astro.new/minimal@next)
- Locally upgrade to the `astro@next` preview
    ```shell
    npm i astro@next
    # If using Framework renderers...
    npm i @astrojs/renderer-name@next
    ```

If you think we made a mistake, **please reopen this issue**!

If you run into any other problems with `astro@next`, please [open a new issue](https://github.com/snowpackjs/astro/issues/new?assignees=&labels=0.21&template=---01-bug-report.yml&title=%F0%9F%90%9B+BUG%3Aastro@next).
